### PR TITLE
Improving API performance

### PR DESCRIPTION
**Pull request by superalex**

### CODE REVIEW

File: taiga/projects/milestones/serializers.py [hunk 1]
```diff
@@ -42,11 +42,11 @@ def get_total_points(self, obj):
...
 
-    def get_client_increment_points(self, obj):
-        return sum(obj.client_increment_points.values())
+    #def get_client_increment_points(self, obj):
+    #    return sum(obj.client_increment_points.values())
 
-    def get_team_increment_points(self, obj):
-        return sum(obj.team_increment_points.values())
+    #def get_team_increment_points(self, obj):
+    #    return sum(obj.team_increment_points.values())
```
Comment (bameda) [inline taiga/projects/milestones/serializers.py]: Remove commented lines

File: taiga/timeline/api.py [hunk 2]
```diff
@@ -46,6 +47,13 @@ def response_for_queryset(self, queryset):
         # Switch between paginated or standard style responses
         page = self.paginate_queryset(queryset)
         if page is not None:
+            user_ids = list(set([obj.data["user"]["id"] for obj in page.object_list]))
```
Comment (bameda) [inline taiga/timeline/api.py]: I think you don't need to make a set and list. 

```
User.objects.filter(id__in=[1, 1, 2, 3])}
```
is equal to
```
User.objects.filter(id__in=[1, 2, 3])}
```

And I would change ```obj.data["user"]["id"]``` to ```obj.data.get("user", {}).get("id", None)```

File: taiga/timeline/api.py [hunk 3]
```diff
@@ -46,6 +47,13 @@ def response_for_queryset(self, queryset):
...
+            User = apps.get_model("users", "User")
+            users = {u.id: u for u in User.objects.filter(id__in=user_ids)}
+
+            for obj in page.object_list:
+                obj._prefetched_user = users.get(obj.data["user"]["id"], None)
```
Comment (bameda) [inline taiga/timeline/api.py]: Be careful with ```obj.data["user"]["id"]```

File: taiga/users/models.py [hunk 4]
```diff
@@ -152,6 +159,64 @@ class Meta:
...
+            for like in self.likes.select_related("content_type").all():
+                like_id = "{}-{}".format(like.content_type.id, like.object_id)
+                self._cached_liked_ids.add(like_id)
+
+        obj_type = apps.get_model("contenttypes", "ContentType").objects.get_for_model(obj)
+        obj_id = "{}-{}".format(obj_type.id, obj.id)
+        return obj_id in self._cached_liked_ids
+
+    def is_watcher(self, obj):
+        ContentType = apps.get_model("contenttypes", "ContentType")
```
Comment (bameda) [inline taiga/users/models.py]: Remove this line, ContentType is imported in the line 27

File: taiga/projects/tasks/api.py [hunk 5]
```diff
@@ -88,6 +88,14 @@ def update(self, request, *args, **kwargs):
     def get_queryset(self):
         qs = super().get_queryset()
         qs = self.attach_votes_attrs_to_queryset(qs)
+        qs = qs.select_related(
+            "milestone",
+            "owner",
+            "assigned_to",
+            "status",
+            "project",
+
```
Comment (bameda) [inline taiga/projects/tasks/api.py]: Remove this line break

File: taiga/projects/milestones/serializers.py [hunk 6]
```diff
@@ -42,11 +42,11 @@ def get_total_points(self, obj):
     def get_closed_points(self, obj):
         return sum(obj.closed_points.values())
 
-    def get_client_increment_points(self, obj):
-        return sum(obj.client_increment_points.values())
+    #def get_client_increment_points(self, obj):
+    #    return sum(obj.client_increment_points.values())
```
Comment (bameda) [inline taiga/projects/milestones/serializers.py]: Remove commented lines

File: taiga/projects/milestones/serializers.py [hunk 7]
```diff
@@ -21,16 +21,16 @@
...
 
 class MilestoneSerializer(WatchersValidator, WatchedResourceModelSerializer, serializers.ModelSerializer):
-    user_stories = UserStorySerializer(many=True, required=False, read_only=True)
+    user_stories = MilestoneUserStorySerializer(many=True, required=False, read_only=True)
     total_points = serializers.SerializerMethodField("get_total_points")
     closed_points = serializers.SerializerMethodField("get_closed_points")
-    client_increment_points = serializers.SerializerMethodField("get_client_increment_points")
-    team_increment_points = serializers.SerializerMethodField("get_team_increment_points")
+    #client_increment_points = serializers.SerializerMethodField("get_client_increment_points")
+    #team_increment_points = serializers.SerializerMethodField("get_team_increment_points")
```
Comment (bameda) [inline taiga/projects/milestones/serializers.py]: Remove commented lines

File: taiga/projects/milestones/serializers.py (hunk 1 above)
Comment (superalex) [inline taiga/projects/milestones/serializers.py]: Ok!

File: taiga/timeline/api.py (hunk 2 above)
Comment (superalex) [inline taiga/timeline/api.py]: About the set and list: it's equal but the finally generated query is more simple so I prefer removing the duplicates, I cant change it anyway if you think it's better.

About the obj data: ok!

File: taiga/timeline/api.py (hunk 3 above)
Comment (superalex) [inline taiga/timeline/api.py]: Ok!

File: taiga/users/models.py (hunk 4 above)
Comment (superalex) [inline taiga/users/models.py]: Ok!

File: taiga/projects/tasks/api.py (hunk 5 above)
Comment (superalex) [inline taiga/projects/tasks/api.py]: Ok!

**Comment by superalex** [comment]
Thank you very very much @bameda I've done all the changes you suggested (except the list/set one) :dancer:

**Comment by bameda** [comment]
Ok with the list/set.

I'm going to merge and rebase in our branche https://github.com/taigaio/taiga-back/pull/530


Gooood work :star2:

**Comment by superalex** [comment]
Thanks!  :racehorse: :racehorse: :racehorse: :racehorse:

**Comment by bameda** [comment]
uohhh I like this horse!

**Comment by superalex** [comment]
It's a race horse so fast as Taiga will be when including this pull request! :D

**Pull request merged**